validator/sawtooth_validator/journal/completer.py

```diff
@@ -39,90 +43,267 @@ class Completer(object):
+            batch_id_list = [x.header_signature for x in block.batches]
+            # Check to see if batchs are in the correct order.
+            if batch_id_list == list(block.header.batch_ids):
+                return block
+            # Check to see if the block has all batch_ids and they can be put
+            # in the correct order
+            elif sorted(batch_id_list) == sorted(list(block.header.batch_ids)):
+                batches = self._finalize_batch_list(block, temp_batches)
+                # Clear batches from block
+                del block.batches[:]
+                # reset batches with full list batches
+                if batches is None:
+                    block.batches.extend(batches)
+                return block
+            else:
+                LOGGER.debug("Block has extra batches. Dropping %s", block)
+                return
+
+    def _finalize_batch_list(self, block, temp_batches):
+        batches = []
+        for batch_id in block.header.batch_ids:
+            if batch_id in self.batch_cache:
+                batches.append(self.batch_cache[batch_id])
+            elif batch_id in temp_batches:
+                batches.append(temp_batches[batch_id])
+            else:
+                return None
+
+        return batches
+
+    def _complete_batch(self, batch):
+        valid = True
+        dependencies = []
+        for txn in batch.transactions:
+            txn_header = TransactionHeader()
+            txn_header.ParseFromString(txn.header)
+            for dependency in txn_header.dependencies:
+                if dependency not in self._seen_txns and not \
+                        self.block_cache.block_store.has_transaction(
+                        dependency):
+                    LOGGER.debug("Transaction %s in batch %s has "
+                                 "unsatisfied dependency: %s",
+                                 txn.header_signature,
+                                 batch.header_signature,
+                                 dependency)
+
+                    dependencies.append(dependency)
+                    if dependency not in self._incomplete_batches:
+                        self._incomplete_batches[dependency] = [batch]
+                    elif batch not in self._incomplete_batches[dependency]:
+                        self._incomplete_batches[dependency] += [batch]
+                    valid = False
+        if not valid:
+            self.gossip.broadcast_batch_by_transaction_id_request(
+                dependencies)
 
         return valid
 
+    def _add_seen_txns(self, batch):
+        for txn in batch.transactions:
+            if txn.header_signature in self._seen_txns and \
+                    self._seen_txns[txn.header_signature] == \
+                    batch.header_signature:
+                break
+            self._seen_txns[txn.header_signature] = batch.header_signature
+            # If there was a batch waiting on this transaction, process that
```

Comment:
this seems like a side effect of add_. Consider renaming the method to provide clarity in calling code.